src/components/color_picker/color_picker.ts

```diff
@@ -232,64 +255,139 @@ export class ColorPicker extends Component<ColorPickerProps, SpreadsheetChildEnv
-    if (!isColorValid(this.state.currentColor)) {
-      ev.stopPropagation();
-      this.state.isCurrentColorInvalid = true;
-      return;
-    }
-    const color = toHex(this.state.currentColor);
+  private setCustomHue(x: Pixel) {
+    // needs to be capped such that h is in [0°, 359°]
+    const h = Math.round(clip((360 * x) / INNER_GRADIENT_WIDTH, 0, 359));
+    this.updateColor({ h });
+  }
+
+  private updateColor(newHsl: Partial<Omit<HSLA, "a">>) {
     this.state.isCurrentColorInvalid = false;
-    this.props.onColorPicked(color);
-    this.state.currentColor = color;
+    this.state.currentHslaColor = { ...this.state.currentHslaColor, ...newHsl };
+    this.state.customHexColor = hslaToHex(this.state.currentHslaColor);
+  }
+
+  onColorClick(color: Color) {
+    if (color) {
+      this.props.onColorPicked(toHex(color));
+    }
+  }
+
+  resetColor() {
+    this.props.onColorPicked("");
   }
 
   toggleColorPicker() {
     this.state.showGradient = !this.state.showGradient;
   }
 
-  computeCustomColor(ev: MouseEvent) {
-    this.state.isCurrentColorInvalid = false;
-    this.state.currentColor = computeCustomColor(ev);
-  }
+  dragGradientPointer(ev: MouseEvent) {
+    const initialGradientCoordinates = { x: ev.offsetX, y: ev.offsetY };
+    this.setCustomGradient(initialGradientCoordinates);
 
-  hideMagnifier(_ev: MouseEvent) {
-    this.state.style.display = "none";
+    const initialMousePosition = { x: ev.clientX, y: ev.clientY };
+
+    const onMouseMove = (ev: MouseEvent) => {
+      const currentMousePosition = { x: ev.clientX, y: ev.clientY };
+      const deltaX = currentMousePosition.x - initialMousePosition.x;
+      const deltaY = currentMousePosition.y - initialMousePosition.y;
+
+      const currentGradientCoordinates = {
+        x: initialGradientCoordinates.x + deltaX,
+        y: initialGradientCoordinates.y + deltaY,
+      };
+      this.setCustomGradient(currentGradientCoordinates);
+    };
+
+    startDnd(onMouseMove, () => {});
   }
 
-  showMagnifier(_ev: MouseEvent) {
-    this.state.style.display = "block";
+  dragHuePointer(ev: MouseEvent) {
+    const initialX = ev.offsetX;
+    const initialMouseX = ev.clientX;
+    this.setCustomHue(initialX);
+    const onMouseMove = (ev: MouseEvent) => {
+      const currentMouseX = ev.clientX;
+      const deltaX = currentMouseX - initialMouseX;
+      const x = initialX + deltaX;
+      this.setCustomHue(x);
+    };
+    startDnd(onMouseMove, () => {});
   }
 
-  moveMagnifier(ev: MouseEvent) {
-    this.state.style.background = computeCustomColor(ev);
-    const shiftFromCursor = 10;
-    this.state.style.left = `${ev.offsetX + shiftFromCursor}px`;
-    this.state.style.top = `${ev.offsetY + shiftFromCursor}px`;
+  setHexColor() {
+    const val = (this.hexInputRef.el as HTMLInputElement)?.value;
+    this.state.customHexColor = val;
+    this.state.isCurrentColorInvalid = false;
+    if (!isColorValid(val)) {
+      this.state.isCurrentColorInvalid = true;
```

Comment:
does `isCurrentColorInvalid` needs to be in a state ? Can't it be computed ?
Also, I'd rename it to `isHexColorInvalid`

Reply:
> does isCurrentColorInvalid needs to be in a state ? Can't it be computed ?
the state is not necessary indeed, i'll remove it


> Also, I'd rename it to `isHexColorInvalid`

The user can technically write `rgb(1,1,1)` and it will be converted via `toHex` helper, so it's kind of a type `Color`  in a large sense. 
I think the helper does this for historical reasons where we did not "force" hex representation for the colors - which we still don't really do. But I think we could explore that possibility in the future, it could simplify some of the color helpers that we have.